lib/services/parser.go

```diff
@@ -518,3 +522,129 @@ func newParserForIdentifierSubcondition(ctx RuleContext, identifier string) (pre
 		},
 	})
 }
+
+// NewResourceParser returns a parser made for boolean expressions based on a
+// json-serialiable resource. Customized to allow short identifiers common in all
+// resources:
+//  - `metadata.name` can be referenced with `name` ie: `name == "jenkins"``
+//  - `metadata.labels + spec.dynamic_labels` can be referenced with `labels`
+//     ie: `labels.env == "prod"`
+// All other fields can be referenced by starting expression with identifier `resource`
+// followed by the names of the json fields ie: `resource.spec.public_addr`.
+func NewResourceParser(resource types.ResourceWithLabels) (BoolPredicateParser, error) {
+	p, err := predicate.NewParser(predicate.Def{
+		Operators: predicate.Operators{
+			AND: predicate.And,
+			OR:  predicate.Or,
+			NOT: predicate.Not,
+			EQ: func(a interface{}, b interface{}) predicate.BoolPredicate {
+				return func() bool {
+					switch aval := a.(type) {
+					case label:
+						bval, ok := b.(string)
+						return ok && aval.value == bval
+					case string:
+						bval, ok := b.(string)
+						return ok && aval == bval
+					default:
+						return false
+					}
+				}
+			},
+			NEQ: func(a interface{}, b interface{}) predicate.BoolPredicate {
+				return func() bool {
+					switch aval := a.(type) {
+					case label:
+						bval, ok := b.(string)
+						return ok && aval.value != bval
+					case string:
+						bval, ok := b.(string)
+						return ok && aval != bval
+					default:
+						return false
+					}
+				}
+			},
+		},
+		Functions: map[string]interface{}{
+			"equals": func(a interface{}, b interface{}) predicate.BoolPredicate {
+				aval, ok := a.(label)
+				if !ok {
+					return predicate.Equals(a, b)
+				}
+				bval, ok := b.(string)
+				return func() bool {
+					return ok && aval.value == bval
+				}
+			},
+			// search allows fuzzy matching against select field values.
+			"search": func(searchVals ...string) predicate.BoolPredicate {
+				return func() bool {
+					return resource.MatchSearch(searchVals)
+				}
+			},
+			// exists checks for an existence of a label by checking
+			// if a key exists. Label value are unchecked.
+			"exists": func(l label) predicate.BoolPredicate {
+				return func() bool {
+					return len(l.key) > 0
+				}
+			},
+		},
+		GetIdentifier: func(fields []string) (interface{}, error) {
+			switch fields[0] {
```

Comment:
What if `len(fields) <= 0`?

Reply:
this will not happen b/c no fields and no operand are errors, i updated test to reflect more edge cases

Reply:
This is a guarantee of the predicate library?

Reply:
yes, it goes through like switch statements to build expressions, and returns errors as it encounters misfits